File: elixir/protein-translation/protein_translation.exs
defmodule ProteinTranslation do

  @codons %{
    "UGU" => "Cysteine",
    "UGC" => "Cysteine",
    "UUA" => "Leucine",
    "UUG" => "Leucine",
    "AUG" => "Methionine",
    "UUU" => "Phenylalanine",
    "UUC" => "Phenylalanine",
    "UCU" => "Serine",
    "UCC" => "Serine",
    "UCA" => "Serine",
    "UCG" => "Serine",
    "UGG" => "Tryptophan",
    "UAU" => "Tyrosine",
    "UAC" => "Tyrosine",
    "UAA" => "STOP",
    "UAG" => "STOP",
    "UGA" => "STOP",
  }

  def of_rna(rna) do
    proteins = get_proteins(String.graphemes(rna))
    cond do
      Enum.member?(proteins, "error") -> {:error, "invalid RNA"}
      true -> {:ok, proteins}
    end
  end

  defp get_proteins([]), do: []
  defp get_proteins([a, b, c | tail]) do
    protein = @codons[a <> b <> c]
    case protein do
      nil -> ["error"]
      "STOP" -> []
      p -> [p | get_proteins(tail)]
    end
  end

  def of_codon(codon) do
    protein = @codons[codon]
    case protein do
      nil -> {:error, "invalid codon"}
      _ -> {:ok, protein}
    end
  end
end

File: python/protein-translation/protein_translation.py
pairs = [
    (['AUG'], 'Methionine'),
    (['UGG'], 'Tryptophan'),
    (['UUA', 'UUG'], 'Leucine'),
    (['UAU', 'UAC'], 'Tyrosine'),
    (['UGU', 'UGC'], 'Cysteine'),
    (['UUU', 'UUC'], 'Phenylalanine'),
    (['UCU', 'UCC', 'UCA', 'UCG'], 'Serine'),
]


def proteins(strand):
    codons = [strand[i:i+3] for i in range(0, len(strand), 3)]
    res = []
    for codon in codons:
        if codon in ['UAA', 'UAG', 'UGA']:
            break
        pair = [p for c, p in pairs if codon in c]
        if len(pair) > 0:
            res.append(pair[0])
    return res
